feature_engine/timeseries/forecasting/window_features.py

```diff
@@ -0,0 +1,333 @@
+                "The dataframe's index contains NaN values or missing data. "
+                "Only dataframes with complete indexes are compatible with "
+                "this transformer."
+            )
+
+        # Check that the index contains unique values.
+        if X.index.is_unique is False:
+            raise NotImplementedError(
+                "The dataframe's index does not contain unique values. "
+                "Only dataframes with unique values in the index are compatible "
+                "with this transformer."
+            )
+
+        # find variables that will be transformed
+        self.variables_ = _find_or_check_numerical_variables(X, self.variables)
+
+        # check if dataset contains na
+        if self.missing_values == "raise":
+            _check_contains_na(X, self.variables_)
+            _check_contains_inf(X, self.variables_)
+
+        self.feature_names_in_ = X.columns.tolist()
+        self.n_features_in_ = X.shape[1]
+
+        return self
+
+    def transform(self, X: pd.DataFrame) -> pd.DataFrame:
+        """
+        Adds window features.
+
+        Parameters
+        ----------
+        X: pandas dataframe of shape = [n_samples, n_features]
+            The data to transform.
+
+        Returns
+        -------
+        X_new: Pandas dataframe, shape = [n_samples, n_features + lag_features]
+            The dataframe with the original plus the new variables.
+        """
+        # checmk method fit has been called
+        check_is_fitted(self)
+
+        # check if 'X' is a dataframe
+        X = _is_dataframe(X)
+
+        # check if input data contains the same number of columns as the fitted dataframe.
+        _check_input_matches_training_df(X, self.n_features_in_)
+
+        # Dataframes must have unique values in the index and no missing data.
+        # Otherwise, when we merge the lag features we will duplicate rows.
+        if X.index.isnull().sum() > 0:
+            raise NotImplementedError(
+                "The dataframe's index contains NaN values or missing data. "
+                "Only dataframes with complete indexes are compatible with "
+                "this transformer."
+            )
+
+        if X.index.is_unique is False:
+            raise NotImplementedError(
+                "The dataframe's index does not contain unique values. "
+                "Only dataframes with unique values in the index are compatible "
+                "with this transformer."
+            )
+
+        # check if dataset contains na
+        if self.missing_values == "raise":
+            _check_contains_na(X, self.variables_)
+            _check_contains_inf(X, self.variables_)
+
+        if self.sort_index is True:
+            X.sort_index(inplace=True)
+
+        # if freq is not None, it overrides periods.
+        if self.freq is not None:
+            tmp = (X[self.variables_]
+                   .rolling(window=self.window).apply(self.function)
```

Comment:
@KishManani

the param windows takes strings and integers. Is this enough? should we consider anything else?

Reply:
I think it covers most use cases. However, in Pandas there's also the option to provide a custom class to define windows (see Pandas rolling docs (here)[BaseIndexer subclass](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.rolling.html]). If the effort required to support it is high, it may not be worth it and can be added in the future.

Reply:
This is what I've done for this parameter:

- removed all init checks. We leave that to pandas. Like this, this param can take whatever pandas accepts. 
- in addition, I also allow passing a a list with window values, so that several window features can be calculated with one transformer.
- in theory, the user could pass a list of callables, but I have not tested this process. @KishManani do you have an example of a useful feature that could be created with a bespoke function? we can use this in a test and more importantly in the documentation.

Reply:
I don't have a good example to hand. My recommendation is to implement it only if an issue is raised with a good use case.